### PR TITLE
Support different cache implementations

### PR DESCRIPTION
**Pull request by thrawn01**
### Purpose
To support alternative hot and main cache implementations that may have better performance under high concurrency workloads. 

### Implementation
* Introduced the `Cache` interface
* Moved the `Modifications from original library` section to the bottom of the readme.
* Renamed `cache` to `mutexCache` and renamed some of it's internal variables
* Modified `mutexCache` to implement the new `Cache` interface
* Moved eviction for `mutexCache` from `populateCache()` into `mutexCache.Add()`
* Added `Options.CacheFactory` to allow the initialization of third party cache implementations
* Added an optional cache implementation for `github.com/maypok86/otter`
* Added `CacheStats.Rejected` to count the number of items rejected by the cache.
* Add documentation for using `otter` in groupcache.
* Renamed `SpawnServer() and `SpawnDaemon()` to `ListenAndServe()`

### Usage
```go
import "github.com/groupcache/groupcache-go/v3/contrib"

// Create a new groupcache instance with a custom cache implementation
instance := groupcache.New(groupcache.Options{
    CacheFactory: func(maxBytes int64) (groupcache.Cache, error) {
        return contrib.NewOtterCache(maxBytes)
    },
    HashFn:    fnv1.HashBytes64,
    Logger:    slog.Default(),
    Transport: t,
    Replicas:  50,
})
```
See #1

### CODE REVIEW

**Comment by thrawn01** [comment]
This is ready for review. @udhos, @gedw99 @Tochemey @Jvb182 @Baliedge @MatthewEdge

I'm also pinging @maypok86 to thank him for the wonderful otter cache! (I hope I'm using it correctly 😄 )

Please review and let me know what I missed or broke.

File: transport/http_transport.go [hunk 1]
```diff
@@ -75,15 +75,15 @@ type Transport interface {
...
+	// ListenAndServe spawns a server that will handle incoming requests for this transport
 	// This is used by daemon and cluster packages to create a cluster of instances using
 	// this specific transport.
-	SpawnServer(ctx context.Context, address string) error
+	ListenAndServe(ctx context.Context, address string) error
 
-	// ShutdownServer shuts down the server started when calling SpawnServer()
-	ShutdownServer(ctx context.Context) error
+	// Shutdown shuts down the server started when calling ListenAndServe()
+	Shutdown(ctx context.Context) error
```
Comment (Tochemey) [inline transport/http_transport.go:84]: @thrawn01 do we have some shutdown timeout(graceful shutdown) concept here or it is not necessary?

**Review by Tochemey — APPROVED** [review]

**Review by Tochemey — COMMENTED** [review]
@thrawn01 I believe like the discovery the contrib cache should go into another repository because of the different cache integration developers may want to add.

**Comment by Tochemey** [comment]
@thrawn01 Also for the cache rejected stats, is it a standard practice out there to reject item? If not I think then we are leaking otter custom feature here.

**Comment by maypok86** [comment]
> @thrawn01 Also for the cache rejected stats, is it a standard practice out there to reject item? If not I think then we are leaking otter custom feature here.

You can, it's your decision. I will even be a little glad, since I was not ready for the popularity of the otter. But I would like to tell you a little about the rejection. There are several ways in which this can happen in different caches.

1. The ristretto approach. We simply discard the inserted item if a high contention is detected. This is a very rare approach and very controversial. To be honest, I really don't like it.
2. Rejection of an item that is too large. This is quite common, because the insertion of such an item will lead to a very strong drop in the hit rate. This is used in otter.

File: transport/http_transport.go (hunk 1 above)
Comment (thrawn01) [reply transport/http_transport.go:84]: Are you talking about the `context.Context` ?

Yes graceful shutdown's can sometimes take a while. (if there is a handler that is stuck for instance) In such a case it's always best to provide a way to timeout a shutdown. I've been bit by this before in production where a service got stuck in shutdown and we had to forceably shutdown the service losing all of the data in the internal buffers that should have been sync'd to disk. We started adding `context.Context` to all our shutdown methods after that. It's one of those things that you don't think is an issue, until it is. 

Golang's standard http server `Shutdown()` method takes a `context.Context` for this same reason.

**Pull request merged**